### ISSUE TITLE
Suggestion: Signals/Observables instead of useWatch$ and useResource$

### ISSUE THREAD

**Issue by mb21** [BuilderIO/qwik#1264]
### Is your feature request related to a problem?

I'm looking at Qwik pretty much the first time now. So take this as feedback as a first impression of the developer API surface of Qwik. I really like the MPA and reactive approach, as well as JSX.

Reposting the [example from the docs](https://qwik.builder.io/docs/components/state/#usewatch-example):

```tsx
const App = component$(() => {
  const state = useStore({
    count: 0,
  });

  const doubleCount = useResource$(({track}) => {
    track(state, 'count');
    return state.count * 2;
  });

  return (
    <>
      <button onClick={() => state.count++}>Increment</button>)}
      Count: {state.count}
      Count * 2: {doubleCount.promise}
    </>
  );
});
```

But I immediately had a bunch of questions:

- why/how does `state.count++` cause an update of the component? Are we overloading the `++` operator or is this transpiled at compile-time by the optimizer? How would this work with non-primitive types like objects or arrays?
- `state.count` is actually a getter function but looks like an ordinary field access?

I could have accepted those, but the `useWatch$` and `useResource$` have really quite a weird API:

- Why are we destructuring `{track}`? Are there other properties passed? which ones?
- What side-effect does the `track` function have that we need to call it? Shouldn't we just pass the observable/signal we want to track as an argument to `useWatch`/`useResource`?
- Can the magic string `'count'` be typechecked against the `state` of the component?

### Describe the solution you'd like

So I immediately thought of a nicer API for `useResource$`, like:

```tsx
const doubleCount = useResource$(count => {
  return count * 2;
}, state.count);
```

But reminding myself of what seemed like magic when seeting the `useStore`, and taking a page from [Solid.js](https://www.solidjs.com/docs/latest#createsignal), maybe even nicer would be something like:

```tsx
const App = component$(() => {
  const [count, setCount] = useStore(0); // or useSignal

  const doubleCount = () => {
    return count() * 2;
  };

  return (
    <>
      <button onClick={() => setCount(count() + 1)}>Increment</button>)}
      Count: {count()}
      Count * 2: {doubleCount()}
    </>
  );
});
```

### Describe alternatives you've considered

(Alternative already mentioned in solution section.)

### Additional context

Again, treat this as early feedback, not knowing about the implementation or other trade-offs you have to consider.

**Comment by manucorporat** [comment]
Unfortunately some of the ideas here cant' be applied because qwik needs to be resumable.



>why/how does state.count++ cause an update of the component? Are we overloading the ++ operator or is this transpiled at compile-time by the optimizer? How would this work with non-primitive types like objects or arrays?

Just a proxy, operators can not be overloader `++` is just syntax sugar for:
```
state.count = state.count + 1;
```

state.count is actually a getter function but looks like an ordinary field access?
yes, it's a getter, using Proxy
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy


```
const doubleCount = useResource$(count => {
  return count * 2;
}, state.count);
```
^ this would not work, the subscription to state.count changing would be done by the render function, not by the useResource()


> Why are we destructuring {track}? Are there other properties passed? which ones?
it's passing a ctx, you can also do: `ctx.track()` typescript will autocomplete it for you
So far, `track`, `cleanup` and `previous`


>What side-effect does the track function have that we need to call it? Shouldn't we just pass the observable/signal we want to track as an argument to useWatch/useResource?

>Can the magic string 'count' be typechecked against the state of the component?
yes, it's already statically typed







`useResource()` in addition solves a completely different problem, even solid has a similar api

**Comment by manucorporat** [comment]
We are planning to implement the signal API but, it's unlikely it replaces useWatch() and useResource()

**Comment by mb21** [comment]
Thanks for the quick reply! Not sure if the API can be made more ergonomical (i.e. declarative) within the constraints of resumability. But if not, consider my questions as something the tutorial or docs could answer somewhere.

---

> We are planning to implement the signal API but, it's unlikely it replaces useWatch() and useResource()

oh, really? I thought signals/observables were just a more declarative interface to what useWatch is doing?

> even solid has a similar api

are you referring to `createResource`? which "creates a signal that reflects the result of an async request". So it's a utility function that returns a signal, not an alternative to signals?

**Comment by manucorporat** [comment]
yes, createResource() is like useResource() in the sense that allows to react to changes in a signal an memoize a async computation, allowing in the meantime to know the status and render the UI accordinly. Going to move this issue into a discussion. Right now we are kind of busy with beta, but i wanna spend some time with this topic before final. 

I think we can improve some of the coherency of the API, regarding useWatch() and useResource() passing a context, instead of the track function directly, ot's because we have amazing plans for this API to extend it with some functionality.
